src/abbreviate.js

```diff
@@ -1,29 +1,70 @@
-import {format} from "d3-format";
+import {formatLocale} from "d3-format";
+import defaultLocale from "./locale.json";
+
+const round = (x, n) => 
+  parseFloat(Math.round(x * Math.pow(10, n)) / Math.pow(10, n)).toFixed(n);
+
+/** */
+function formatSuffix(value, precision, suffixes) {
+  let i = 0;
+  if (value) {
+    if (value < 0) value *= -1;
+    i = 1 + Math.floor(1e-12 + Math.log(value) / Math.LN10);
+    i = Math.max(-24, Math.min(24, Math.floor((i - 1) / 3) * 3));
+  }
+  const d = suffixes[8 + i / 3];
+  
+  return {
+    number: round(d.scale(value), precision),
+    symbol: d.symbol
+  };
+}
+
+/** */
+function parseSuffixes(d, i) {
+  const k = Math.pow(10, Math.abs(8 - i) * 3);
+  return {
+    scale: i > 8 ? d => d / k : d => d * k,
+    symbol: d
+  };
+}
+
 
 /**
     @function formatAbbreviate
     @desc Formats a number to an appropriate number of decimal places and rounding, adding suffixes if applicable (ie. `1200000` to `"1.2M"`).
-    @param {Number} n The number to be formatted.
+    @param {Number|String} n The number to be formatted.
+    @param {Object|String} locale The locale config to be used. If *value* is an object, the function will format the numbers according the object. The object must include `suffixes`, `delimiter` and `currency` properties.
     @returns {String}
 */
-export default function(n) {
-  if (typeof n !== "number") return "N/A";
-  const length = n.toString().split(".")[0].replace("-", "").length;
+export default function(n, locale = "en-US") {
+  if (isFinite(n)) n *= 1;
+  else return "N/A";
+
+  const length = n.toString().split(".")[0].replace("-", "").length,
+        localeConfig = typeof locale === "object" ? locale : defaultLocale[locale] || defaultLocale["en-US"],
+        suffixes = localeConfig.suffixes.map(parseSuffixes);
+
+  const separator = localeConfig.separator || "";
+
+  const d3plusFormatLocale = formatLocale({
+    currency: localeConfig.currency || ["$", ""],
+    decimal: localeConfig.delimiters.decimal || ".",
+    grouping: localeConfig.grouping || [3],
+    thousands: localeConfig.delimiters.thousands || ","
+  });
+
   let val;
   if (n === 0) val = "0";
   else if (length >= 3) {
-    const f = format(".3s")(n)
-      .replace("G", "B")
-      .replace("T", "t")
-      .replace("P", "q")
-      .replace("E", "Q");
-    const num = f.slice(0, -1);
-    const char = f.slice(f.length - 1);
-    val = `${parseFloat(num)}${char}`;
+    const f = formatSuffix(d3plusFormatLocale.format(".3r")(n), 2, suffixes);
```

Comment:
Now you're using BOTH the d3 number formatter and your own! Shouldn't the final logic be something more like this?

```js
val = `${d3plusFormatLocale.format(".3r")(n)}${separator}${char}`;
```

Reply:
@davelandry `d3plusFormatLocale.format(".3r")(n)` only returns the number in decimal notation, rounded to significant digits.

```
d3plusFormatLocale.format(".3r")(123456) -> 123000
```

and `formatSuffix` returns the suffix and the first characters of the number

Reply:
Ahh ok interesting, I guess I didn't understand how the format locale works (the other d3 format functions usually return an abbreviated string). Get a few more test in there for delimiters and we should be good to go!